### PR TITLE
fix scheduler version skew handling for in-place pod resize

### LINKED ISSUE #117767: [FG:InPlacePodVerticalScaling] Implement version skew handling for in-place pod resize
### What would you like to be added?

This is a tracking bug to implement two changes to better handle version skew as documented in https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#version-skew-strategy
1. kubelet: When feature-gate is disabled, if kubelet sees a Proposed resize, it rejects the resize as Infeasible.
2. scheduler: If PodStatus.Resize field is not empty, scheduler uses max(ResourcesAllocated, Requests) even if the feature-gate is disabled.

### Why is this needed?

This is needed to gracefully handle component version skews that can occur during upgrades.

### PR DESCRIPTION
**Pull request by wlp1153468871**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

If PodStatus.Resize field is not empty, scheduler uses max(ResourcesAllocated, Requests) even if the feature-gate is disabled. 

#### Which issue(s) this PR fixes:

Fixes #117767 

#### Does this PR introduce a user-facing change?

```release-note
 NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#version-skew-strategy
```

### CODE REVIEW

File: pkg/api/v1/resource/helpers_test.go [hunk 1]
```diff
@@ -720,7 +720,7 @@ func TestPodResourceRequests(t *testing.T) {
 		{
 			description: "resized, infeasible",
 			expectedRequests: v1.ResourceList{
-				v1.ResourceCPU: resource.MustParse("2"),
```
Comment (AxeZhan) [inline pkg/api/v1/resource/helpers_test.go:723]: Why are we changing these resource size in tests?🤔️

Comment (wlp1153468871) [reply pkg/api/v1/resource/helpers_test.go:723]: According to the chapter `Version Skew Strategy` of the KEP-1287, when  kube-scheduler computes the pod resource requests,  it should use max(AllocatedResources, Requests) if PodStatus.Resize != "".

In this test case, the PodStatus.Resize is equal to Infeasible and the AllocatedResources is less than Requests, so the expected result is Requests. I'm not sure that this change is reasonable when the PodStatus.Resize is equal to Infeasible. If I misunderstood, please correct me. Thanks.

FYI:

- https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#version-skew-strategy
- https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L2576

Comment (AxeZhan) [reply pkg/api/v1/resource/helpers_test.go:723]: Ah I see, sgtm

**Comment by wlp1153468871** [comment]
@vinaykul  Can you help me review a PR? Thank you

**Comment by alexzielenski** [comment]
untagged mistakenly

/sig api-machinery
/triage accepted
/cc @deads2k

**Comment by wlp1153468871** [comment]
@mikedanese @sttts Can you help me review a PR? Thank you

**Comment by vinaykul** [comment]
@wlp1153468871 Thank you very much for picking this up! Could we please fix both kubelet and scheduler fix in a single PR so that it fully resolves the issue?

We should add an e2e test for this.. maybe we can extend test/e2e/node/pod_resize.go 
In order to exercise `InPlacePodVerticalScaling` feature-gate enabled in apiserver but disabled in kubelet and scheduler, one way might be to clone [this CI job](https://github.com/kubernetes/test-infra/blob/dd7bacb796d5956ae9e1874e55ba1482d505a029/config/jobs/kubernetes/sig-node/containerd.yaml#L1391L1428) into two other CI jobs (maybe with 72 hour cadence) and set this FG to false using --feature-gates flag passed via KUBELET_TEST_ARGS and SCHEDULER_TEST_ARGS.

@bobbypage @BenTheElder Does that above suggestion look reasonable? (Will setting the --feature-gates flag override for kubelet & scheduler? If not, would the inverse work?)

I'm also wondering if an e2e skew testing CI job already exists. It sounds a boilerplate version skew test job for our version skew support matrix is something many alpha and beta features could use, and in general to detect skew regressions. I do see references to `--check-version-skew` flag in test-infra repo but my git grep shows them all set to false.

**Comment by wlp1153468871** [comment]
> @wlp1153468871 Thank you very much for picking this up! Could we please fix both kubelet and scheduler fix in a single PR so that it fully resolves the issue?
> 
> We should add an e2e test for this.. maybe we can extend test/e2e/node/pod_resize.go In order to exercise `InPlacePodVerticalScaling` feature-gate enabled in apiserver but disabled in kubelet and scheduler, one way might be to clone [this CI job](https://github.com/kubernetes/test-infra/blob/dd7bacb796d5956ae9e1874e55ba1482d505a029/config/jobs/kubernetes/sig-node/containerd.yaml#L1391L1428) into two other CI jobs (maybe with 72 hour cadence) and set this FG to false using --feature-gates flag passed via KUBELET_TEST_ARGS and SCHEDULER_TEST_ARGS.
> 
> @bobbypage @BenTheElder Does that above suggestion look reasonable? (Will setting the --feature-gates flag override for kubelet & scheduler? If not, would the inverse work?)
> 
> I'm also wondering if an e2e skew testing CI job already exists. It sounds a boilerplate version skew test job for our version skew support matrix is something many alpha and beta features could use, and in general to detect skew regressions. I do see references to `--check-version-skew` flag in test-infra repo but my git grep shows them all set to false.

Thank you very much for your reply. However, I have noticed that pbialon has already claimed the task related to kubelet. I would be grateful to hear your thoughts on this matter, @bobbypage @BenTheElder

**Comment by vinaykul** [comment]
> Thank you very much for your reply. However, I have noticed that pbialon has already claimed the task related to kubelet. I would be grateful to hear your thoughts on this matter, @bobbypage @BenTheElder

That's ok, we can have two PRs. The core code change is quite simple in this but the idea is for new contributors to get a broader view of the upgrade scenario considerations and how it can be exercised in CI so it two contributors get to learn, twice as better.

**Pull request closed without merging**